Userland/Libraries/LibCore/Resource.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023, Andrew Kaster <akaster@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/ByteBuffer.h>
+#include <AK/Error.h>
+#include <AK/MemoryStream.h>
+#include <AK/OwnPtr.h>
+#include <AK/Span.h>
+#include <AK/String.h>
+#include <AK/StringView.h>
+#include <AK/Variant.h>
+#include <LibCore/File.h>
+#include <LibCore/MappedFile.h>
+
+namespace Core {
+class Resource {
+public:
+    static ErrorOr<NonnullOwnPtr<Resource>> load_from_uri(StringView);
+
+    [[nodiscard]] bool is_file() const { return !m_data.has<DirectoryTag>(); }
+    [[nodiscard]] bool is_directory() const { return m_data.has<DirectoryTag>(); }
+
+    [[nodiscard]] String uri() const;
+    [[nodiscard]] String filename() const;
+    [[nodiscard]] Optional<String> filesystem_path() const;
+
+    [[nodiscard]] ByteBuffer clone_data() const;
+    [[nodiscard]] ByteBuffer release_data() &&;
+    [[nodiscard]] ReadonlyBytes data() const;
+    [[nodiscard]] FixedMemoryStream stream() const;
+
+    [[nodiscard]] Vector<String> children() const;
+    // Depth-first
+    template<IteratorFunction<Resource const&> Callback>
+    IterationDecision for_each_descendant(Callback&&) const;
+
+    template<IteratorFunction<Resource const&> Callback>
+    void for_each_descendant_file(Callback&&) const;
+
+    struct DirectoryTag { };
+
+private:
+    friend class ResourceImplementation;
+
+    enum class Scheme {
+        File,
+        Resource,
+    };
+
+    Resource(String path, Scheme, NonnullOwnPtr<Core::MappedFile>);
+    Resource(String path, Scheme, ByteBuffer);
+    Resource(String path, Scheme, DirectoryTag);
+
+    String m_path; // Relative to scheme root. File: abspath, Resource: resource root
```

Comment:
LexicalPath

Reply:
In theory, no one should be inspecting this as a filesystem path, so I don't think it makes sense to go whole LexicalPath on it. it's just a unique identifier for a resource.